File: mobile/src/main/java/net/mm2d/android/cds/Tag.java
/*
 * Copyright (c) 2016 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.android.cds;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

/**
 * CdsObjectのタグ情報を表現するクラス
 *
 * Elementのままでは情報の参照コストが高いため、
 * よりシンプルな構造に格納するためのクラス。
 * CdsObjectのXMLはタグが入れ子になることはないため
 * タグ＋値、属性＋値の情報を表現できれば十分。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
public class Tag implements Parcelable {
    private final String mName;
    private final String mValue;
    private final Map<String, String> mAttribute;

    /**
     * インスタンス作成。
     *
     * パッケージ外でのインスタンス化禁止
     *
     * @param element タグ情報
     */
    Tag(@NonNull Element element) {
        this(element, false);
    }

    /**
     * インスタンス作成。
     *
     * パッケージ外でのインスタンス化禁止
     *
     * @param element タグ情報
     * @param root    タグがitem/containerのときtrue
     */
    Tag(@NonNull Element element, boolean root) {
        this(element, root ? "" : element.getTextContent());
    }

    /**
     * インスタンス作成。
     *
     * @param element タグ情報
     * @param value   タグの値
     */
    private Tag(@NonNull Element element, @NonNull String value) {
        mName = element.getTagName();
        mValue = value;
        final NamedNodeMap attributes = element.getAttributes();
        final int size = attributes.getLength();
        if (size == 0) {
            mAttribute = Collections.emptyMap();
        } else {
            mAttribute = new LinkedHashMap<>(size);
        }
        for (int i = 0; i < size; i++) {
            final Node attr = attributes.item(i);
            mAttribute.put(attr.getNodeName(), attr.getNodeValue());
        }
    }

    /**
     * タグ名を返す。
     *
     * @return タグ名
     */
    @NonNull
    public String getName() {
        return mName;
    }

    /**
     * タグの値を返す。
     *
     * @return タグの値
     */
    @NonNull
    public String getValue() {
        return mValue;
    }

    /**
     * 属性値を返す。
     *
     * @param name 属性名
     * @return 属性値、見つからない場合null
     */
    @Nullable
    public String getAttribute(@Nullable String name) {
        return mAttribute.get(name);
    }

    /**
     * 属性値を格納したMapを返す。
     *
     * @return 属性値を格納したUnmodifiable Map
     */
    @NonNull
    public Map<String, String> getAttributes() {
        if (mAttribute.size() == 0) {
            return Collections.emptyMap();
        }
        return Collections.unmodifiableMap(mAttribute);
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();
        sb.append(mValue);
        for (final Entry<String, String> entry : mAttribute.entrySet()) {
            sb.append("\n");
            sb.append("@");
            sb.append(entry.getKey());
            sb.append(" => ");
            sb.append(entry.getValue());
        }
        return sb.toString();
    }

    /**
     * Parcelable用のコンストラクタ。
     *
     * @param in Parcel
     */
    private Tag(@NonNull Parcel in) {
        mName = in.readString();
        mValue = in.readString();
        final int size = in.readInt();
        if (size == 0) {
            mAttribute = Collections.emptyMap();
        } else {
            mAttribute = new LinkedHashMap<>(size);
            for (int i = 0; i < size; i++) {
                final String name = in.readString();
                final String value = in.readString();
                mAttribute.put(name, value);
            }
        }
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(mName);
        dest.writeString(mValue);
        dest.writeInt(mAttribute.size());
        for (final Entry<String, String> entry : mAttribute.entrySet()) {
            dest.writeString(entry.getKey());
            dest.writeString(entry.getValue());
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    /**
     * Parcelableのためのフィールド
     */
    public static final Creator<Tag> CREATOR = new Creator<Tag>() {
        @Override
        public Tag createFromParcel(Parcel in) {
            return new Tag(in);
        }

        @Override
        public Tag[] newArray(int size) {
            return new Tag[size];
        }
    };
}

File: mobile/src/main/java/net/mm2d/android/upnp/cds/Tag.kt
/*
 * Copyright (c) 2016 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.android.upnp.cds

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import net.mm2d.upnp.util.asIterable
import org.w3c.dom.Element
import java.util.*

/**
 * シンプルなXMLのタグ情報を表現するクラス
 *
 * Elementのままでは情報の参照コストが高いため、
 * よりシンプルな構造に格納するためのクラス。
 * CdsObjectのXMLのようにElementが入れ子になることのない
 * タグ＋値、属性＋値の情報を表現できれば十分なものを表現するのに使用する。
 * 入れ子関係を持つXMLは表現できない。
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class Tag(
    val name: String,
    val value: String,
    val attributes: Map<String, String>
) : Parcelable {
    /**
     * 属性値を返す。
     *
     * @param name 属性名
     * @return 属性値、見つからない場合null
     */
    fun getAttribute(name: String?): String? {
        return attributes[name]
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append(value)
        attributes.forEach {
            sb.append("\n@${it.key} => ${it.value}")
        }
        return sb.toString()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeString(value)
        dest.writeInt(attributes.size)
        attributes.forEach {
            dest.writeString(it.key)
            dest.writeString(it.value)
        }
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Creator<Tag> {
        val EMPTY = Tag("", "", emptyMap())
        override fun createFromParcel(parcel: Parcel): Tag = create(parcel)
        override fun newArray(size: Int): Array<Tag?> = arrayOfNulls(size)

        private fun create(parcel: Parcel): Tag {
            val name = parcel.readString()!!
            val value = parcel.readString()!!
            val size = parcel.readInt()
            val map = LinkedHashMap<String, String>(size)
            for (i in 0 until size) {
                map[parcel.readString()!!] = parcel.readString()!!
            }
            return Tag(name, value, map)
        }

        /**
         * インスタンス作成。
         *
         * パッケージ外でのインスタンス化禁止
         *
         * @param element タグ情報
         * @param root    タグがitem/containerのときtrue
         */
        @JvmOverloads
        fun create(element: Element, root: Boolean = false): Tag =
            create(element, if (root) "" else element.textContent)

        /**
         * インスタンス作成。
         *
         * @param element タグ情報
         * @param value   タグの値
         */
        private fun create(element: Element, value: String): Tag = Tag(
            element.tagName,
            value,
            element.attributes.asIterable().map { it.nodeName to it.nodeValue }.toMap()
        )
    }
}
